File: 20210724_Saturday/20210724.rb
# Leetcode: 126. Word Ladder II.
# this code does not works because need back tracking.
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {String[][]}
def find_ladders(begin_word, end_word, word_list)
  @curr_path = []
  @shortest_paths = []
  @adj_list = {}

  # Coping the words into the set of efficient deletion in BFS.
  copied_word_list = Set.new(word_list)
  # Build the DAG using BFS.
  bfs(begin_word, end_word, copied_word_list)

  # Every path will start from the beging_word.
  @curr_path.push(begin_word)
  # Treverse the DAG to find all the paths between begin_word and end_word.
  backtrack(begin_word,end_word)

  return @shortest_paths
end

def bfs(begin_word, end_word, word_list)
  q = []
  q.push(begin_word)

  # remove the root word wich is the first layer in the BFS.
  word_list.delete?(begin_word)

  isEnqueued = {}
  isEnqueued[begin_word] = true
  while !q.empty?
    visited = []
    q.size.times do
      curr_word = q.shift
      # findNeighbor will have the adjacent words of the curr_word.
      neighbors = findNeighbors(curr_word, word_list)
      neighbors.each do |word|
        visited.push[word]
        @adj_list[curr_word] ||= []
        @adj_list[curr_word].push(word)
        if isEnqueued[word].nil?
          q.push(word)
          isEnqueued[word] = true
        end
      end # neightbors
    end # q.size
    # removing the words of the previous layer.
    visited.each{|wrd| word_list.delete?(wrd)}
  end # while
end

File: 20210724_Saturday/20210724.java
// https://leetcode.com/problems/word-ladder-ii/solution/
class Solution {
    Map<String, List<String>> adjList = new HashMap<String, List<String>>();
    List<String> currPath = new ArrayList<String>();
    List<List<String>> shortestPaths = new ArrayList<List<String>>();
    
    private List<String> findNeighbors(String word, Set<String> wordList) {
        List<String> neighbors = new ArrayList<String>();
        char charList[] = word.toCharArray();
        
        for (int i = 0; i < word.length(); i++) {
            char oldChar = charList[i];   
            
            // replace the i-th character with all letters from a to z except the original character
            for (char c = 'a'; c <= 'z'; c++) {
                charList[i] = c;
                
                // skip if the character is same as original or if the word is not present in the wordList
                if (c == oldChar || !wordList.contains(String.valueOf(charList))) {
                    continue;
                }
                neighbors.add(String.valueOf(charList));
            }
            charList[i] = oldChar;
        }
        return neighbors;
    }
    
    private void backtrack(String source, String destination) {
        // store the path if we reached the endWord
        if (source.equals(destination)) {
            List<String> tempPath = new ArrayList<String>(currPath);
            shortestPaths.add(tempPath);
        }
        
        if (!adjList.containsKey(source)) {
            return;
        }
        
        for (int i = 0; i < adjList.get(source).size(); i++) {
            currPath.add(adjList.get(source).get(i));
            backtrack(adjList.get(source).get(i), destination);
            currPath.remove(currPath.size() - 1);
        }
    }
    
    private void bfs(String beginWord, String endWord, Set<String> wordList) {
        Queue<String> q = new LinkedList<>();
        q.add(beginWord);
        
        // remove the root word which is the first layer in the BFS
        if (wordList.contains(beginWord)) {
            wordList.remove(beginWord);
        }
        
        Map<String, Integer> isEnqueued = new HashMap<String, Integer>();
        isEnqueued.put(beginWord, 1);
        
        while (q.size() > 0)  {
             // visited will store the words of current layer
            List<String> visited = new ArrayList<String>();
            
            for (int i = q.size() - 1; i >= 0; i--) {
                String currWord = q.peek(); 
                q.remove();

                // findNeighbors will have the adjacent words of the currWord
                List<String> neighbors = findNeighbors(currWord, wordList);
                for (String word : neighbors) {
                    visited.add(word);
                    
                    if (!adjList.containsKey(currWord)) {
                        adjList.put(currWord, new ArrayList<String>());
                    }
                    
                    // add the edge from currWord to word in the list
                    adjList.get(currWord).add(word);
                    if (!isEnqueued.containsKey(word)) {
                        q.add(word);
                        isEnqueued.put(word, 1);
                    }
                }
            }
             // removing the words of the previous layer
            for (int i = 0; i < visited.size(); i++) {
                if (wordList.contains(visited.get(i))) {
                    wordList.remove(visited.get(i));
                }
            }
        }
    }
    
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        // copying the words into the set for efficient deletion in BFS
        Set<String> copiedWordList = new HashSet<>(wordList);
        // build the DAG using BFS
        bfs(beginWord, endWord, copiedWordList);
        
        // every path will start from the beginWord
        currPath.add(beginWord);
        // traverse the DAG to find all the paths between beginWord and endWord
        backtrack(beginWord, endWord);
        
        return shortestPaths;
    }
}
